packages/wonder-stuff-i18n/src/utils/localize-file.js

```diff
@@ -0,0 +1,143 @@
+ * the new file hashes for JS files, if they've changed.
+ *
+ * @param {string} locale the locale to localize to
+ * @param {string} content the JS file contents to localize
+ * @param {Object} hashMap a mapping of old to new hashes
+ */
+export const localizeString = ({
+    locale,
+    content,
+    hashMap,
+    getLocalePath,
+}: {
+    locale: string,
+    content: string,
+    hashMap?: ?{[string]: string},
+    getLocalePath: (string) => string,
+}): string => {
+    // Replace all English paths with the path to the locale
+    const newContent = content.replace(
+        new RegExp(getLocalePath("en"), "g"),
+        getLocalePath(locale),
+    );
+
+    if (!hashMap) {
+        return newContent;
+    }
+
+    // Look for all hash-like strings and replace the ones that can be found
+    // in the hash map, leaving all others intact
+    return newContent.replace(HASH_RE, (hash) => hashMap[hash] || hash);
+};
+
+/**
+ * Produce an MD5 hash of the contents of a file that is 20 characters long.
+ *
+ * @param {string} contents The contents of a file
+ */
+export const hashFileContents = (contents: string): string =>
+    crypto.createHash("md5").update(contents).digest("hex").slice(0, 20);
+
+/**
+ * Localizes a file with the new public path for the specific locale and with
+ * the new file hashes for JS files, if they've changed.
+ *
+ * NOTE: hashMap is mutated when this is called, it's updated with any new
+ * hashes and their mapping.
+ *
+ * @param {string} locale the locale to localize to
+ * @param {string} fileName the file name to read in and out to localize
+ * @param {Object} hashMap a mapping of old to new hashes, this object is
+ * updated when new hashes are generated for files.
+ */
+export const localizeFile = ({
+    content,
+    locale,
+    oldHash,
+    hashMap,
+    getLocalePath,
+}: {
+    content: string,
+    locale: string,
+    oldHash: ?string,
+    hashMap: {[string]: string},
+    getLocalePath: (string) => string,
+}): string => {
+    const newContent = localizeString({
+        locale,
+        content,
+        hashMap,
+        getLocalePath,
+    });
+
+    if (oldHash != null) {
+        // Compute the new hash of the file contents
+        const newHash = hashFileContents(newContent);
+
+        // If the hash of the file contents has changed then we need to
+        // update the hash in the file name and also update it in other
+        // sources (such as the manifest files), so we update the hashMap
+        // so that it will update when we call localizeFile on the manifests.
+        if (oldHash !== newHash) {
+            hashMap[oldHash] = newHash;
+
+            // We also need to update any existing mappings that already are
+            // in place to point to the new hash as well
+            for (const hash of Object.keys(hashMap)) {
+                if (hashMap[hash] === oldHash) {
+                    hashMap[hash] = newHash;
+                }
+            }
```

Comment:
I'm not sure what this security check is actually testing:
```
> const a = {};
undefined
> a
{}
> a[proto] = {foo: () => console.log("foo")}
Uncaught ReferenceError: proto is not defined
> 
```

Reply:
Hmm? I'm not sure I follow... we're inserting a property whose value is stored in hash.